// hark-master/index.js

'use strict';

"use strict";

/* ========================================== wildemitter ======================================== */



window.WildEmitter=function() { 

module.exports = WildEmitter;

WildEmitter.mixin = function (constructor) {
    var prototype = constructor.prototype || constructor;

    prototype.isWildEmitter= true;

    // Listen on the given `event` with `fn`. Store a group name if present.
    prototype.on = function (event, groupName, fn) {
        this.callbacks = this.callbacks || {};
        var hasGroup = (arguments.length === 3),
            group = hasGroup ? arguments[1] : undefined,
            func = hasGroup ? arguments[2] : arguments[1];
        func._groupName = group;
        (this.callbacks[event] = this.callbacks[event] || []).push(func);
        return this;
    };

    // Adds an `event` listener that will be invoked a single
    // time then automatically removed.
    prototype.once = function (event, groupName, fn) {
        var self = this,
            hasGroup = (arguments.length === 3),
            group = hasGroup ? arguments[1] : undefined,
            func = hasGroup ? arguments[2] : arguments[1];
        function on() {
            self.off(event, on);
            func.apply(this, arguments);
        }
        this.on(event, group, on);
        return this;
    };

    // Unbinds an entire group
    prototype.releaseGroup = function (groupName) {
        this.callbacks = this.callbacks || {};
        var item, i, len, handlers;
        for (item in this.callbacks) {
            handlers = this.callbacks[item];
            for (i = 0, len = handlers.length; i < len; i++) {
                if (handlers[i]._groupName === groupName) {
                    //console.log('removing');
                    // remove it and shorten the array we're looping through
                    handlers.splice(i, 1);
                    i--;
                    len--;
                }
            }
        }
        return this;
    };

    // Remove the given callback for `event` or all
    // registered callbacks.
    prototype.off = function (event, fn) {
        this.callbacks = this.callbacks || {};
        var callbacks = this.callbacks[event],
            i;

        if (!callbacks) return this;

        // remove all handlers
        if (arguments.length === 1) {
            delete this.callbacks[event];
            return this;
        }

        // remove specific handler
        i = callbacks.indexOf(fn);
        callbacks.splice(i, 1);
        if (callbacks.length === 0) {
            delete this.callbacks[event];
        }
        return this;
    };

    /// Emit `event` with the given args.
    // also calls any `*` handlers
    prototype.emit = function (event) {
        this.callbacks = this.callbacks || {};
        var args = [].slice.call(arguments, 1),
            callbacks = this.callbacks[event],
            specialCallbacks = this.getWildcardCallbacks(event),
            i,
            len,
            item,
            listeners;

        if (callbacks) {
            listeners = callbacks.slice();
            for (i = 0, len = listeners.length; i < len; ++i) {
                if (!listeners[i]) {
                    break;
                }
                listeners[i].apply(this, args);
            }
        }

        if (specialCallbacks) {
            len = specialCallbacks.length;
            listeners = specialCallbacks.slice();
            for (i = 0, len = listeners.length; i < len; ++i) {
                if (!listeners[i]) {
                    break;
                }
                listeners[i].apply(this, [event].concat(args));
            }
        }

        return this;
    };

    // Helper for for finding special wildcard event handlers that match the event
    prototype.getWildcardCallbacks = function (eventName) {
        this.callbacks = this.callbacks || {};
        var item,
            split,
            result = [];

        for (item in this.callbacks) {
            split = item.split('*');
            if (item === '*' || (split.length === 2 && eventName.slice(0, split[0].length) === split[0])) {
                result = result.concat(this.callbacks[item]);
            }
        }
        return result;
    };

};

WildEmitter.mixin(WildEmitter);
}

/* ================================= hark ======================================= */




window.Hark=function() {
    
var WildEmitter = new WildEmitter();

function getMaxVolume (analyser, fftBins) {
  var maxVolume = -Infinity;
  analyser.getFloatFrequencyData(fftBins);

  for(var i=4, ii=fftBins.length; i < ii; i++) {
    if (fftBins[i] > maxVolume && fftBins[i] < 0) {
      maxVolume = fftBins[i];
    }
  };

  return maxVolume;
}


var audioContextType;
if (typeof window !== 'undefined') {
  audioContextType = window.AudioContext || window.webkitAudioContext;
}
// use a single audio context due to hardware limits
var audioContext = null;
module.exports = function(stream, options) {
  var harker = new WildEmitter();


  // make it not break in non-supported browsers
  if (!audioContextType) return harker;

  //Config
  var options = options || {},
      smoothing = (options.smoothing || 0.1),
      interval = (options.interval || 50),
      threshold = options.threshold,
      play = options.play,
      history = options.history || 10,
      running = true;

  //Setup Audio Context
  if (!audioContext) {
    audioContext = new audioContextType();
  }
  var sourceNode, fftBins, analyser;

  analyser = audioContext.createAnalyser();
  analyser.fftSize = 512;
  analyser.smoothingTimeConstant = smoothing;
  fftBins = new Float32Array(analyser.frequencyBinCount);

  if (stream.jquery) stream = stream[0];
  if (stream instanceof HTMLAudioElement || stream instanceof HTMLVideoElement) {
    //Audio Tag
    sourceNode = audioContext.createMediaElementSource(stream);
    if (typeof play === 'undefined') play = true;
    threshold = threshold || -50;
  } else {
    //WebRTC Stream
    sourceNode = audioContext.createMediaStreamSource(stream);
    threshold = threshold || -50;
  }

  sourceNode.connect(analyser);
  if (play) analyser.connect(audioContext.destination);

  harker.speaking = false;

  harker.setThreshold = function(t) {
    threshold = t;
  };

  harker.setInterval = function(i) {
    interval = i;
  };

  harker.stop = function() {
    running = false;
    harker.emit('volume_change', -100, threshold);
    if (harker.speaking) {
      harker.speaking = false;
      harker.emit('stopped_speaking');
    }
    analyser.disconnect();
    sourceNode.disconnect();
  };
  harker.speakingHistory = [];
  for (var i = 0; i < history; i++) {
      harker.speakingHistory.push(0);
  }

  // Poll the analyser node to determine if speaking
  // and emit events if changed
  var looper = function() {
    setTimeout(function() {

      //check if stop has been called
      if(!running) {
        return;
      }

      var currentVolume = getMaxVolume(analyser, fftBins);

      harker.emit('volume_change', currentVolume, threshold);

      var history = 0;
      if (currentVolume > threshold && !harker.speaking) {
        // trigger quickly, short history
        for (var i = harker.speakingHistory.length - 3; i < harker.speakingHistory.length; i++) {
          history += harker.speakingHistory[i];
        }
        if (history >= 2) {
          harker.speaking = true;
          harker.emit('speaking');
        }
      } else if (currentVolume < threshold && harker.speaking) {
        for (var i = 0; i < harker.speakingHistory.length; i++) {
          history += harker.speakingHistory[i];
        }
        if (history == 0) {
          harker.speaking = false;
          harker.emit('stopped_speaking');
        }
      }
      harker.speakingHistory.shift();
      harker.speakingHistory.push(0 + (currentVolume > threshold));

      looper();
    }, interval);
  };
  looper();


  return harker;
}

}




/* ============================================ actions ======================================= */




var hark =new Hark();


var iraVolumes = [];
var tagVolumes = [];
var streamVolumes = [];
var referenceVolumes = [];
for (var i = 0; i < 100; i++) {
  iraVolumes.push(-100);
  tagVolumes.push(-100);
  streamVolumes.push(-100);
  referenceVolumes.push(-50);
}

//(function() {
//  //Video Tag Demo
//  var stream = document.querySelector('#ira video');
//  var speechEvents = hark(stream);
//  var notification = document.querySelector('#iraSpeaking');
//  var log = bows('Video Demo');
//
//  speechEvents.on('speaking', function() {
//    log('speaking');
//    notification.style.display = 'block';
//  });
//
//  speechEvents.on('volume_change', function(volume, threshold) {
//    //log('volume change', volume, threshold);
//    iraVolumes.push(volume);
//    iraVolumes.shift();
//  });
//
//  speechEvents.on('stopped_speaking', function() {
//    log('stopped_speaking');
//    notification.style.display = 'none';
//  });
//})();


(function() {
  //Audio Tag Demo
  var stream = document.querySelector('audio');
  var speechEvents = hark(stream);
  var notification = document.querySelector('#UserSpeakingf');

  speechEvents.on('speaking', function() {
    log('speaking');
    alert('speaking');
    notification.style.display = 'block';
  });

  speechEvents.on('volume_change', function(volume, threshold) {
    //log('volume change', volume, threshold);
    tagVolumes.push(volume);
    tagVolumes.shift();
  });

  speechEvents.on('stopped_speaking', function() {
    log('stopped_speaking');
    alert('stopped_speaking');
    notification.style.display = 'none';
  });
})();


//(function() {
//  //Microphone demo
//  var getUserMedia = require('getusermedia');
//  var attachmediastream = require('attachmediastream');
//  var notification = document.querySelector('#userSpeaking');
//  var log = bows('Microphone Demo');
//
//  getUserMedia(function(err, stream) {
//    if (err) throw err
//
//    attachmediastream(stream, document.querySelector('#mic video'));
//    var speechEvents = hark(stream);
//
//    speechEvents.on('speaking', function() {
//      notification.style.display = 'block';
//      log('speaking');
//    });
//
//    speechEvents.on('volume_change', function(volume, threshold) {
//      log(volume, threshold)
//      streamVolumes.push(volume);
//      streamVolumes.shift();
//    });
//
//    speechEvents.on('stopped_speaking', function() {
//      notification.style.display = 'none';
//      log('stopped_speaking');
//    });
//  });
//})();
//
//(function () {
//
//  function drawLine(canvas, data, color) {
//    var drawContext = canvas.getContext('2d');
//    drawContext.moveTo(0,canvas.height);
//    drawContext.beginPath();
//    drawContext.strokeStyle = color;
//    for (var i = 0; i < data.length; i++) {
//      var value = -data[i];
//      var percent = value / 100;
//      var height = canvas.height * percent;
//      var vOffset = height; //canvas.height - height - 5;
//      var hOffset = i * canvas.width / 100.0;
//      drawContext.lineTo(hOffset, vOffset);
//    }
//    drawContext.stroke();
//  }
//  function draw() {
//    var canvas = document.querySelector('canvas');
//    if (!canvas) return;
//    var drawContext = canvas.getContext('2d');
//    drawContext.clearRect (0, 0, canvas.width, canvas.height);
//
//    drawLine(canvas, iraVolumes, 'red');
//    drawLine(canvas, tagVolumes, 'green');
//    drawLine(canvas, streamVolumes, 'blue');
//    drawLine(canvas, referenceVolumes, 'black');
//    window.requestAnimationFrame(draw);
//  }
//  window.requestAnimationFrame(draw);
//})();


module.exports = function (stream, el, options) {
    var URL = window.URL;
    var opts = {
        autoplay: true,
        mirror: false,
        muted: false
    };
    var element = el || document.createElement('video');
    var item;

    if (options) {
        for (item in options) {
            opts[item] = options[item];
        }
    }

    if (opts.autoplay) element.autoplay = 'autoplay';
    if (opts.muted) element.muted = true;
    if (opts.mirror) {
        ['', 'moz', 'webkit', 'o', 'ms'].forEach(function (prefix) {
            var styleName = prefix ? prefix + 'Transform' : 'transform';
            element.style[styleName] = 'scaleX(-1)';
        });
    }

    // this first one should work most everywhere now
    // but we have a few fallbacks just in case.
    if (URL && URL.createObjectURL) {
        element.src = URL.createObjectURL(stream);
    } else if (element.srcObject) {
        element.srcObject = stream;
    } else if (element.mozSrcObject) {
        element.mozSrcObject = stream;
    } else {
        return false;
    }

    return element;
};
